### ISSUE TITLE
get_transient_version called multiple times on checkout

### ISSUE THREAD

**Issue by glagonikas** [woocommerce/woocommerce#17632]
## Prerequisites
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue
1. Update woocommerce\includes\class-wc-cache-helper.php Line 123 and append `error_log($transient_value . " - " . $refresh);`
2. Disable all plugins leaving WC and WC Stripe active
3. Enable storefront
4. Update wp-config.php to allow loging to debug.log
5. Process an order on incognito as a new customer buying a simple product

## Expected behavior
See only one record on the debug.log like
```
[08-Nov-2017 14:50:38 UTC] 1510152638 - 1
```

## Actual behavior
The debug.log is showing three records like
```
[08-Nov-2017 14:50:38 UTC] 1510152638 - 1
[08-Nov-2017 14:50:43 UTC] 1510152643 - 1
[08-Nov-2017 14:50:45 UTC] 1510152645 - 1
```

### 1. Affects other plugins
When WC Subscriptions are active an additional log is added even if you buy a simple product.

I understand this is probably to do with WCS, but it could also be an indication that other official plugins are unintentionally affected by this.

### 2. Cause of Deadlocks
I understand this function is crucial, however we also need to consider its impact on the server load and the overall processing speed when an order is placed.

Each time the function is called, the code will do a number of queries to the database including DELETE and UPDATE.

On an Innodb table, this can cause (and has been causing us) deadlocks. Those might go unnoticed in smaller sites but they become an issue on bigger sites.

Additionally, those deadlocks are preventing other plugins from firing correctly. In our case, they are preventing Subscriptions from being created. - I have raised this with the relevant team asking for a workaround from their end, but haven't heard back still.

### 3. Slow checkout
Since the update to WCv3 the checkout has become unreasonably slow and i suspect it's partly due to the number of queries made by this function.

## Isolating the problem
- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment
<details>
```
### WordPress Environment ###

Home URL: https://www.site.com
Site URL: https://www.site.com
WC Version: 3.2.3
Log Directory Writable: ✔
WP Version: 4.8.3
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: ✔
WP Cron: –
Language: en_GB

### Server Environment ###

Server Info: Apache
PHP Version: 7.0.25
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.55.1
OpenSSL/1.0.2m

SUHOSIN Installed: –
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.2.3
WC Database Prefix: wp_
Total Database Size: 181.20MB
Database Data Size: 113.07MB
Database Index Size: 68.13MB
wp_woocommerce_sessions: Data: 0.19MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_order_itemmeta: Data: 0.30MB + Index: 0.09MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB
wp_comments: Data: 0.05MB + Index: 0.14MB
wp_links: Data: 0.02MB + Index: 0.02MB
wp_options: Data: 3.45MB + Index: 0.28MB
wp_postmeta: Data: 21.55MB + Index: 46.69MB
wp_posts: Data: 73.55MB + Index: 17.95MB
wp_po_plugins: Data: 0.00MB + Index: 0.00MB
wp_sib_export_processes: Data: 0.17MB + Index: 0.00MB
wp_termmeta: Data: 0.02MB + Index: 0.05MB
wp_terms: Data: 0.22MB + Index: 0.25MB
wp_term_relationships: Data: 2.52MB + Index: 1.52MB
wp_term_taxonomy: Data: 0.20MB + Index: 0.23MB
wp_usermeta: Data: 0.23MB + Index: 0.20MB
wp_users: Data: 0.02MB + Index: 0.08MB
wp_woocommerce_coupon_activation_codes: Data: 8.52MB + Index: 0.00MB

### Post Type Counts ###

attachment: 582
page: 169
payment_retry: 8823
post: 1
product: 151
product_variation: 39
revision: 5083
scheduled-action: 18
shop_coupon: 398
shop_order: 25
shop_subscription: 4
wc_order_email: 1
wc_order_status: 31

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (2) ###

WooCommerce Stripe Gateway: by WooCommerce – 3.2.3 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.2.3

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: composite (composite)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: ❌ Page not set
Basket: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/
Terms and conditions: #93 - /terms-conditions/

### Theme ###

Name: Storefront
Version: 2.2.5
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build
personally we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –
```
</details>

**Comment by rodrigoprimo** [comment]
Thanks for your detailed report, @glagonikas.

> Update woocommerce\includes\class-wc-cache-helper.php Line 123

For future reference, below is the link to the affected line (`class-wc-cache-helper.php` was updated on master and line 123 is now a comment):

[class-wc-cache-helper.php#L137](https://github.com/woocommerce/woocommerce/blob/7b9fb6c0cb129e977800d9450d630e6dbe6b6a6e/includes/class-wc-cache-helper.php#L137
)

> Each time the function is called, the code will do a number of queries to the database including DELETE and UPDATE.

It seems to me that only when this method is called under two specific circumstances (called for a `$group` that was never used before or with `$refresh` set to `true`) the code will touch the database. Most of the calls to this method don't touch the database at all.

When `WC_Cache_Helper::get_transient_version()` is called with a new `$group` three queries are executed:

```
wp> WC_Cache_Helper::get_transient_version( 'new-group' );

Executed queries:
SELECT option_value FROM wp_options WHERE option_name = '_transient_timeout_new-group-transient-version' LIMIT 1
SELECT option_value FROM wp_options WHERE option_name = '_transient_new-group-transient-version' LIMIT 1
INSERT INTO `wp_options` (`option_name`, `option_value`, `autoload`) VALUES ('_transient_new-group-transient-version', '1510258222', 'yes') ON DUPLICATE KEY UPDATE `option_name` = VALUES(`option_name`), `option_value` = VALUES(`option_value`), `autoload` = VALUES(`autoload`)
```

When it is called with `$refresh` set to `true` two queries are executed:

```
wp> WC_Cache_Helper::get_transient_version( 'orders', true );

Executed queries:
DELETE FROM wp_options WHERE option_name LIKE '\\_transient\\_%1510258328' ORDER BY option_id LIMIT 1000
UPDATE `wp_options` SET `option_value` = '1510258332' WHERE `option_name` = '_transient_orders-transient-version'
```

I was able to reproduce what is described in this issue. When I placed an order, `WC_Cache_Helper::get_transient_version()` was called three times with the parameter `$refresh` set to `true`, executing three times variations of the two queries shown above. They get called three times because `wc_delete_shop_order_transients()` is called three times during checkout. The first time when the order is created and two subsequent times when Stripe updates the order.

I don't see an easy way to make those queries run only once or to improve the performance of the queries. We could explore removing the `ORDER BY` from the `DELETE` query as I'm not sure if we need it but I'm also not sure if this change would have a significant impact. Do you have any suggestion, @glagonikas?

> On an Innodb table, this can cause (and has been causing us) deadlocks. Those might go unnoticed in smaller sites but they become an issue on bigger sites.

I don't think this issue should be a priority for the core team to work on as big sites using an external cache won't be impacted by this (WP stores transients in memory and not in the database when an external cache is used), but a PR that address this is welcome.

**Comment by glagonikas** [comment]
Hi @rodrigoprimo 

> Each time the function is called, the code will do a number of queries to the database including DELETE and UPDATE.

Apologies for not clarifying the above bit. It is indeed an issue when called with the refresh variable set to true. Looking at the code, this is called by https://github.com/woocommerce/woocommerce/blob/7b9fb6c0cb129e977800d9450d630e6dbe6b6a6e/includes/wc-order-functions.php#L444

Each time an order is updated this function is called, meaning Subscriptions will also call it a couple of times and in our case, will also fire once more as we automatically classify orders and move them to custom statuses. On our setup and with all plugins enabled, it fires 9 times for each checkout.

We're only just exploring external cache (Redis) and we haven't tested on the live site yet. Even though we've got an optimised SQL, i noticed it doesn't take many users purchasing at the same time, before it starts throwing deadlocks, so i'd imagine smaller sites could also be affected.

As stated before, it's probably an issue for applications that:
1. Don't use External Cache
2. Have the wp_options table in InnoDB
3. Have the SQL isolation level set to REPEATABLE-READ (default) (https://mariadb.com/kb/en/library/set-transaction/#isolation-level) - This probably needs confirming from a DB expert, but reading about it, changing it to READ-COMMITTED would probably resolve some of the deadlocks.

With the above in mind, i do agree it's not an urget issue, but would probably make sense to keep in mind and try to find a solution that reduces the number of queries required on each checkout.

My suggestion would be to use the function after all processing has been finished, although i do see an argument for clearing the transients each time an order is changed, so i'm not sure which approach would be best.

Can you see a reason why a deadlock thrown by the function would prevent other plugins from firing correctly? In our case, each time a deadlock occurs, a subscription is not generated for that checkout. Already asked this to the relevant team, but haven't received a response yet. Would it make sense to wrap the function in try-catch to prevent this?

**Comment by rodrigoprimo** [comment]
I'm closing this issue as won't fix because I don't see an easy way around this and I'm assuming that the number of affected installs is small (only high traffic sites without external cache). That being said, I will see if I can check when I have some time why we need to clear all transients (groups) on updates. Maybe it can be conditional.

PRs are welcome in case someone has time to work on this before me.

**Comment by crstauf** [comment]
I'm 95% confident we're having the exact issue described with a client's site hosted on WP Engine, potentially resulting in missing address data, update to order status, and (as OP mentioned) missing subscription objects.

Given the mention that external cache should resolve the issue, I'll check with WPE and make sure our caching rules are setup properly.

**Comment by crstauf** [comment]
So in our case, the size of options that were autloaded was much larger than the server's buffer, so operations were being run directly on the database, thus causing the deadlocks. I have cleaned up some of the autoloaded options (to bring it within the buffer's max size), and will monitor to see if deadlocks continue.

**Comment by crstauf** [comment]
It does not appear that cleaning up the autoloaded options in `wp_options` is sufficient to resolve the deadlocks. Continuing to investigate with WP Engine.

**Comment by willgorham** [comment]
I'm seeing db deadlocks after the 3.4.x update as well. Haven't identified the issue yet, or even isolated it to WC, but figured I'd add my voice here.

**Comment by mikejolley** [comment]
This issue was reported long before 3.4 so coincidental?

cc @kloon

**Comment by willgorham** [comment]
@mikejolley Probably coincidental. We made some caching changes too at that time so I'm still looking into what caused this to surface for us all of a sudden.

**Comment by crstauf** [comment]
The epic continues: we had successfully implemented object caching, but an attempt to export order/customer data (via [plugin](https://woocommerce.com/products/ordercustomer-csv-export/)) caused objects in the object cache to be constantly evicted, resulting in no caching and expensive queries being run on nearly every request.

I will contact SkyVerge and inform them of the issue, but thought it might be relevant here as well, as not all plugins will likely be developed to properly support an environment with object caching, so +1 for those queries during checkout being cleaned up.

**Comment by OxfordDigital** [comment]
We are also seeing many deadlock errors since upgrade from 3.2.6 to 3.4.2. It is affecting orders, and leading them to have status "Cancelled" after the timeout occurs - even though payment is successful. Host is WPEngine. 

[Mon Jun 18 11:38:55.075281 2018] [:error] [pid 19358] [client *.*.*.*:12559] WordPress database error Deadlock found when trying to get lock; try restarting transaction for query DELETE FROM wp_options WHERE option_name LIKE '\\\\_transient\\\\_%152933434' ORDER BY option_id LIMIT 1000; /* From [sitename.co.uk/?wc-ajax=checkout] in [/path/sitename/wp-content/plugins/woocommerce/includes/class-wc-cache-helper.php:166] */ made by require('wp-blog-header.php'), require_once('wp-includes/template-loader.php'), do_action('template_redirect'), WP_Hook->do_action, WP_Hook->apply_filters, call_user_func_array, WC_AJAX::do_wc_ajax, do_action('wc_ajax_checkout'), WP_Hook->do_action, WP_Hook->apply_filters, call_user_func_array, WC_AJAX::checkout, WC_Checkout->process_checkout, WC_Checkout->process_order_payment, WC_Gateway_Stripe->process_payment, WC_Stripe_Payment_Gateway->process_response, WC_Order->payment_complete, WC_Order->save, WC_Data_Store->update, WC_Order_Data_Store_CPT->update, Abstract_WC_Order_Data_Store_CPT->update, Abstract_WC_Order_Data_Store_CPT->clear_caches, wc_delete_shop_order_transients, WC_Cache_Helper::get_transient_version, WC_Cache_Helper::delete_version_transients, referer: https://sitename.co.uk/checkout/

**Comment by crstauf** [comment]
Hey @OxfordDigital, I had success preventing deadlocks by activating object caching, but that can introduce other problems, so consider carefully.

**Comment by rodrigoprimo** [comment]
@crstauf, do you believe the problems you started experiencing when you enabled object caching are related to WooCommerce core? If so, could you please open an issue and copy me? WC core should work flawlessly with object caching enabled. Thanks.

**Comment by glagonikas** [comment]
In our case we experienced some issues with object caching due to our server setup which scales horizontally. It was mainly queries not refreshing correctly and didn't affect orders or subscriptions.

We didn't pinpoint the issues down to WC core and it could be any of our other plugins, so we've temporarily disabled it.

We resolved the deadlocks by disabling delete_version_transients using a filter. We then use a cron job to delete them manually.

I don't think 3.4 alone is to blame. We've always had those deadlocks since upgrading to v3, due to the size and traffic of the website, but they've gradually been getting worse (and the checkout has been getting slower).

**Comment by crstauf** [comment]
@rodrigoprimo Echoing what glagonikas said: can't confirm an issue with core, but some plugins (https://github.com/woocommerce/woocommerce/issues/17632#issuecomment-396962879) don't seem to have been tested in environments with object caching.

For us, deadlocks starting happening around the same time we upgraded to 3.4, but again like glagonikas said, I had noticed deadlocks occur infrequently prior to 3.4, but 3.4 seems to have increased their frequency.

**Comment by crstauf** [comment]
@OxfordDigital The filter is `woocommerce_delete_version_transients_limit`; if you return `0`, the function that is primarily responsible for the deadlocks will no longer run (https://github.com/woocommerce/woocommerce/pull/19894). But you need to setup a cron to clear out the transients, otherwise your `wp_options` table will grow pretty quickly. (I have a question out to glagonikas on what query he runs for the cron job.)

The use of object caching completely disables the offending functionality, thus preventing the deadlock from occurring. (Since object caching evicts objects using LRU algorithm, cleanup isn't necessary).

**Comment by glagonikas** [comment]
We only clear expired transients based on time, so we remove the ones where their expiry date is in the past

**Comment by jessuppi** [comment]
For anyone needing a user-friendly cleanup approach, we made this for our Woo clients:

https://wordpress.org/plugins/delete-expired-transients-littlebizzy/

...it supports the following constants:

```
define('DELETE_EXPIRED_TRANSIENTS_HOURS', '6');
define('DELETE_EXPIRED_TRANSIENTS_MAX_EXECUTION_TIME', '10');
define('DELETE_EXPIRED_TRANSIENTS_MAX_BATCH_RECORDS', '50');
```

**Comment by rodrigoprimo** [comment]
Thanks for your reply.

Not directly related to this issue, but since @OxfordDigital @crstauf and @glagonikas mentioned an increase in deadlocks since upgrading to 3.4, I thought it was worth mentioning PR #20537 in case you haven't checked it already. WC 3.4.3, which was released yesterday, includes this PR which addresses a deadlock issue introduced in 3.4.0.

**Comment by OxfordDigital** [comment]
@crstauf @glagonikas @rodrigoprimo many thanks for all your input on this. We resolved the issue by setting numerous wp_options to autoload=no set by poor plugins, and then enabling the object cache. We will also apply 3.4.3 soon.

**Comment by Protohominid** [comment]
> I'm closing this issue as won't fix because I don't see an easy way around this and I'm assuming that the number of affected installs is small (only high traffic sites without external cache). That being said, I will see if I can check when I have some time why we need to clear all transients (groups) on updates. Maybe it can be conditional.
> 
> PRs are welcome in case someone has time to work on this before me.

We are also seeing these deadlock errors in our php logs. Can you clarify what you mean by 'external cache'?

**Comment by crstauf** [comment]
@Protohominid External cache (also known as object cache) is an enhancement some servers implement to improve performance. Simply put, queries to the database are stored in memory rather than executed each time, which makes the response faster. External cache in WordPress is also used by the `WP_Cache` class, to store transients and options in the external cache.

Note: while options are both stored in the object cache and written to the database, transients are only stored in the object cache.

**Comment by crstauf** [comment]
@Protohominid No... this issue is suspected to exist primarily on high traffic sites _without_ object caching enabled; if you have it on, leave it on.

**Comment by crstauf** [comment]
@Protohominid Here's the article I referenced to get the MySQL output:
https://www.percona.com/blog/2014/10/28/how-to-deal-with-mysql-deadlocks/
